lib/hexpm/repository/release.ex

```diff
@@ -155,6 +155,34 @@ defmodule Hexpm.Repository.Release do
       select: {p.name, r.version, r.inserted_at, p.meta}
     )
   end
+
+  def downloads_by(release_id, filter) do
+    case filter do
+      :by_day ->
+        from(d in Download,
+          where: d.release_id == ^release_id,
+          group_by: fragment("date_trunc('day', ?)", d.day),
+          order_by: fragment("date_trunc('day', ?)", d.day),
+          select: {fragment("to_char(date_trunc('day', ?), 'YYYY-MM-DD')", d.day), fragment("sum(?)", d.downloads)})
+      :by_week ->
+        from(d in Download,
+          where: d.release_id == ^release_id,
+          group_by: fragment("date_trunc('week', ?)",  d.day), # d.day,
+          order_by: fragment("date_trunc('week', ?)",  d.day), # d.day,
+          select: {fragment("to_char(date_trunc('week', ?), 'YYYY-MM-DD')", d.day), fragment("sum(?)", d.downloads)})
+      :by_month ->
+        from(d in Download,
+          where: d.release_id == ^release_id,
+          group_by: fragment("date_trunc('month', ?)",  d.day), # d.day,
+          order_by: fragment("date_trunc('month', ?)",  d.day), # d.day,
+          select: {fragment("to_char(date_trunc('month', ?), 'YYYY-MM')", d.day), fragment("sum(?)", d.downloads)})
```

Comment:
Since these queries only differ in the `select` part you can use composable ecto queries to build everything except select once and then compose the select for each filter.

Reply:
This is the maximum I achieve https://elixirforum.com/t/how-would-you-refactor-a-code-with-fragment-ecto-macros/9191/5. So, it is a bit refactored but not so much, because the select, group_by and order_by depends on the `from d` clause. If you have better refactoring strategy, let me know.

```
  def downloads_by_period(release_id, filter) do
    query = from(d in Download,
      where: d.release_id == ^release_id)
    case filter do
      "day" ->
        from(d in query,
          group_by: date_trunc("day", d.day),
          order_by: date_trunc("day", d.day),
          select: {date_trunc_format("day", "YYYY-MM-DD", d.day), sum(d.downloads)})
      "week" ->
        from(d in query,
          group_by: date_trunc("week", d.day),
          order_by: date_trunc("week", d.day),
          select: {date_trunc_format("week", "YYYY-MM-DD", d.day), sum(d.downloads)})
      "month" ->
        from(d in query,
          group_by: date_trunc("month", d.day),
          order_by: date_trunc("month", d.day),
          select: {date_trunc_format("month", "YYYY-MM", d.day), sum(d.downloads)})
      "all" ->
        from(d in query,
          select: sum(d.downloads))
    end
  end

```

Reply:
It looks like `group_by` and `order_by` are shared by all filters except "all"`. You can special case by matching it on it in the head `def downloads_by_period(release_id, "all") and then handle the rest with a shared query.